sdk/lib/cmlib/cmcheck.c

```diff
@@ -0,0 +1,1800 @@
+    SiblingNode = (PCM_KEY_NODE)HvGetCell(Hive, Sibling);
+    if (!SiblingNode)
+    {
+        /* Couldn't get the sibling node, bail out */
+        DPRINT1("CmpValidateLexicalOrder(): Failed to get the sibling node!\n");
+        return FALSE;
+    }
+
+    /* CASE 1: Two distinct non-compressed Unicode names */
+    if ((ChildNode->Flags & KEY_COMP_NAME) == 0 &&
+        (SiblingNode->Flags & KEY_COMP_NAME) == 0)
+    {
+        Result = CmpCompareDistinctNames(&(SiblingNode->Name[0]),
+                                         SiblingNode->NameLength,
+                                         &(ChildNode->Name[0]),
+                                         ChildNode->NameLength);
+        if (Result >= 0)
+        {
+            DPRINT1("CmpValidateLexicalOrder(): The sibling node name is greater or equal to that of the child!\n");
+            return FALSE;
+        }
+    }
+
+    /* CASE 2: Both compressed Unicode names */
+    if ((ChildNode->Flags & KEY_COMP_NAME) &&
+        (SiblingNode->Flags & KEY_COMP_NAME))
+    {
+        Result = CmpCompareBothCompressedNames(SiblingNode->Name,
+                                               SiblingNode->NameLength,
+                                               ChildNode->Name,
+                                               ChildNode->NameLength);
+        if (Result >= 0)
+        {
+            DPRINT1("CmpValidateLexicalOrder(): The compressed sibling node name is greater or equal to that of the compressed child!\n");
+            return FALSE;
+        }
+    }
+
+    /* CASE 3: The child name is compressed but the sibling is not */
+    if ((ChildNode->Flags & KEY_COMP_NAME) &&
+        (SiblingNode->Flags & KEY_COMP_NAME) == 0)
+    {
+        SiblingString.Buffer = &(SiblingNode->Name[0]);
+        SiblingString.Length = SiblingNode->NameLength;
+        SiblingString.MaximumLength = SiblingNode->NameLength;
+        Result = CmpCompareCompressedName(&SiblingString,
+                                          ChildNode->Name,
+                                          ChildNode->NameLength);
+        if (Result >= 0)
+        {
+            DPRINT1("CmpValidateLexicalOrder(): The sibling node name is greater or equal to that of the compressed child!\n");
+            return FALSE;
+        }
+    }
+
+    /* CASE 4: The sibling name is compressed but the child is not */
+    if ((SiblingNode->Flags & KEY_COMP_NAME) &&
+        (ChildNode->Flags & KEY_COMP_NAME) == 0)
+    {
+        ChildString.Buffer = &(ChildNode->Name[0]);
+        ChildString.Length = ChildNode->NameLength;
+        ChildString.MaximumLength = ChildNode->NameLength;
+        Result = CmpCompareCompressedName(&ChildString,
+                                          SiblingNode->Name,
+                                          SiblingNode->NameLength);
+        if (Result <= 0)
+        {
+            DPRINT1("CmpValidateLexicalOrder(): The compressed sibling node name is lesser or equal to that of the child!\n");
```

Comment:
By grepping your code for !\n and counting the results you will see how many bytes you are wasting. Also long lines make it more likely to generate line-breaks in the logoutput due to limited window size, making the log harder to read. Each logged char consumes time and bytes. I would recommend stripping all your ! at the end of log lines therefore.